torch/sparse/semi_structured.py

```diff
@@ -320,70 +314,25 @@ def __torch_dispatch__(cls, func, types, args, kwargs) -> Any:
-    mask: Optional[torch.Tensor] = None,
-    transposed: bool = False,
-) -> SparseSemiStructuredTensor:
-    """
-    This function converts a dense tensor into a sparse semi-structured tensor.
-    It will return a SparseSemiStructuredTensor, a subclass of torch.Tensor.
-
-    This function will check to ensure the dense tensor has the right dtype, size, dims, and device.
-    We currently only support semi-structured sparse tensors for 2d CUDA tensors.
-    Additionally, your tensor must be a positive multiple of a block size given the dtype
-
-    - torch.float16  (r, c) must be >= and a multiple of 64
-    - torch.int8     (r, c) must be >= and a multiple of 128
-
-    Args:
-        original_tensor (Tensor): the dense tensor to convert
-        mask (Optional BoolTensor): boolean mask to apply to the original tensor
-        transposed (bool, optional): whether the dense tensor is transposed
-
-    Returns:
-        SparseSemiStructuredTensor: A sparse semi-structured tensor created from the given original_tensor and mask
-
-    Raises:
-        None
-
-    Example:
-        >>> # xdoctest: +REQUIRES(env:TORCH_DOCTEST_CUDA)
-        >>> A = torch.Tensor([0, 0, 1, 1]).tile((128, 32)).half().cuda()
-        tensor([[0., 0., 1.,  ..., 0., 1., 1.],
-                [0., 0., 1.,  ..., 0., 1., 1.],
-                [0., 0., 1.,  ..., 0., 1., 1.],
-                ...,
-                [0., 0., 1.,  ..., 0., 1., 1.],
-                [0., 0., 1.,  ..., 0., 1., 1.],
-                [0., 0., 1.,  ..., 0., 1., 1.]], device='cuda:0', dtype=torch.float16)
-        >>> A_sparse = to_sparse_semi_structured(A, mask=A.bool())
-        SparseSemiStructuredTensor(shape=torch.Size([128, 128]), transposed=False, values=tensor([[1., 1., 1.,  ..., 1., 1., 1.],
-                [1., 1., 1.,  ..., 1., 1., 1.],
-                [1., 1., 1.,  ..., 1., 1., 1.],
-                ...,
-                [1., 1., 1.,  ..., 1., 1., 1.],
-                [1., 1., 1.,  ..., 1., 1., 1.],
-                [1., 1., 1.,  ..., 1., 1., 1.]], device='cuda:0', dtype=torch.float16),
-            metadata=tensor([[-4370, -4370, -4370,  ..., -4370, -4370, -4370],
-                [-4370, -4370, -4370,  ..., -4370, -4370, -4370],
-                [-4370, -4370, -4370,  ..., -4370, -4370, -4370],
-                ...,
-                [-4370, -4370, -4370,  ..., -4370, -4370, -4370],
-                [-4370, -4370, -4370,  ..., -4370, -4370, -4370],
-                [-4370, -4370, -4370,  ..., -4370, -4370, -4370]], device='cuda:0',
-       dtype=torch.int16))
-    """
-    return SparseSemiStructuredTensor(original_tensor, mask=mask, transposed=transposed)
+        m, n = self.shape
+        indices_dtype = SparseSemiStructuredTensor.__get_indices_dtype(self.dtype)
+
+        from torch.sparse._semi_structured_conversions import sparse_semi_structured_to_dense
+
+        return sparse_semi_structured_to_dense(
+            self.compressed_tensor[: m * n // 2].view(m, -1),
```

Comment:
nit: can you use `.values` and `.indicies` here instead ?

Reply:
No, compiler doesn't like this either, following error is raised:

<details>
<summary>
Traceback
</summary>

```
Traceback (most recent call last):
  File "/home/alex/scratch/git/torch/sparse/semi_structured.py", line 335, in to_dense
    return sparse_semi_structured_to_dense(
  File "/home/alex/scratch/git/torch/_dynamo/eval_frame.py", line 295, in _fn
    return fn(*args, **kwargs)
  File "/home/alex/scratch/git/torch/_dynamo/eval_frame.py", line 448, in catch_errors
    return callback(frame, cache_size, hooks, frame_state)
  File "/home/alex/scratch/git/torch/_dynamo/convert_frame.py", line 531, in _convert_frame
    result = inner_convert(frame, cache_size, hooks, frame_state)
  File "/home/alex/scratch/git/torch/_dynamo/convert_frame.py", line 127, in _fn
    return fn(*args, **kwargs)
  File "/home/alex/scratch/git/torch/_dynamo/convert_frame.py", line 360, in _convert_frame_assert
    return _compile(
  File "/home/alex/scratch/git/torch/_dynamo/utils.py", line 180, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/alex/scratch/git/torch/_dynamo/convert_frame.py", line 519, in _compile
    raise InternalTorchDynamoError(str(e)).with_traceback(e.__traceback__) from None
  File "/home/alex/scratch/git/torch/_dynamo/convert_frame.py", line 430, in _compile
    out_code = transform_code_object(code, transform)
  File "/home/alex/scratch/git/torch/_dynamo/bytecode_transformation.py", line 1000, in transform_code_object
    transformations(instructions, code_options)
  File "/home/alex/scratch/git/torch/_dynamo/convert_frame.py", line 400, in transform
    tracer = InstructionTranslator(
  File "/home/alex/scratch/git/torch/_dynamo/symbolic_convert.py", line 1979, in __init__
    self.symbolic_locals = collections.OrderedDict(
  File "/home/alex/scratch/git/torch/_dynamo/symbolic_convert.py", line 1982, in <genexpr>
    VariableBuilder(self, LocalSource(k))(f_locals[k]),
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 207, in __call__
    vt = self._wrap(value).clone(**self.options())
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 372, in _wrap
    return type_dispatch(self, value)
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 691, in wrap_listlike
    output = [
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 692, in <listcomp>
    VariableBuilder(self.tx, GetItemSource(self.get_source(), i))(
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 207, in __call__
    vt = self._wrap(value).clone(**self.options())
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 372, in _wrap
    return type_dispatch(self, value)
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 898, in wrap_tensor
    tensor_variable = wrap_fx_proxy(
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 1118, in wrap_fx_proxy
    return wrap_fx_proxy_cls(
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 1179, in wrap_fx_proxy_cls
    example_value = wrap_to_fake_tensor_and_record(
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 1459, in wrap_to_fake_tensor_and_record
    fake_e = wrap_fake_exception(
  File "/home/alex/scratch/git/torch/_dynamo/utils.py", line 891, in wrap_fake_exception
    return fn()
  File "/home/alex/scratch/git/torch/_dynamo/variables/builder.py", line 1460, in <lambda>
    lambda: tx.fake_mode.from_tensor(
  File "/home/alex/scratch/git/torch/_subclasses/fake_tensor.py", line 1587, in from_tensor
    return self.fake_tensor_converter(
  File "/home/alex/scratch/git/torch/_subclasses/fake_tensor.py", line 340, in __call__
    return self.from_real_tensor(
  File "/home/alex/scratch/git/torch/_subclasses/fake_tensor.py", line 293, in from_real_tensor
    out = self.meta_converter(
  File "/home/alex/scratch/git/torch/_subclasses/meta_utils.py", line 535, in __call__
    r = self.meta_tensor(
  File "/home/alex/scratch/git/torch/_subclasses/meta_utils.py", line 295, in meta_tensor
    base = self.meta_tensor(
  File "/home/alex/scratch/git/torch/_subclasses/meta_utils.py", line 414, in meta_tensor
    s = t.untyped_storage()
torch._dynamo.exc.InternalTorchDynamoError: python bindings to nullptr storage (e.g., from torch.Tensor._make_wrapper_subclass) are currently unsafe and thus disabled.  See https://github.com/pytorch/pytorch/issues/61669 for more details
```
</details>

Again, if someone could suggest a solution, please do.

Reply:
I see let's just keep this as is for now. I'll follow up with core and see if this something that will go away when they add compile support for subclassing.